main.js

```diff
@@ -0,0 +1,280 @@
+// const axios = require('axios')
+
+let countries = [];
+let APIerror = "";
+
+// delete a DOMtree, including the parent element in the parameter to this function
+
+function delDomTree(root) {
+    for (let i = 0; i < root.children.length; i++) {
+        delDomTree(root.children[i]);
+    }
+    root.remove();
+}
+
+// if string is undefined, empty or contains less than 4 characters
+// return the empty message
+
+function cleanString(string, emptyMessage) {
+
+    let out = string ?? emptyMessage;
+    if (out.includes('[') ) out = emptyMessage;
+    return out;
+
+}
+
+
+// format an array of objects with the name attribute to a comma
+// separated string with last 2 element separated by 'and' in stead of ','.
+// so results:
+// one -> one
+// one,two -> one and two
+// one,two,thrre -> one,two and three.
+
+function A2String(A) {
+    let string = '';
+    let alen = A.length;
+    for (let i = 0; i < alen; ++i) {
+        string += cleanString(A[i].name, "unspecified others")
+        switch (i) {
+            case (alen - 1):
+                break;
+            case (alen - 2):
+                string += ' and ';
+                break;
+            default:
+                string += ', ';
+                break;
+        }
+    }
+
+    //console.log( string );
+
+    return string;
+}
+
+
+// Collect all countries and select relevant fields for our app
+async function getCountries() {
+    APIerror = "";
+    try {
+        const result = await axios.get('https://restcountries.eu/rest/v2/all?fields=name;capital;population;currencies;languages;flag;subregion');
+
+        for (let i in result.data) {
+            let c = {};
+            ({
+                name: c.name,
+                capital: c.capital,
+                languages: c.languages,
+                currencies: c.currencies,
+                flag: c.flag,
+                population: c.population,
+                subregion: c.subregion
+            } = result.data[i]);
+            countries.push(c);
+        }
+        //console.log( countries );
+    } catch (e) {
+        console.error(e);
+        APIerror = " Unable to collect country data. The network may be failing or the website may be down. Try again later."
+        console.warn(APIerror);
```

Comment:
Eigenlijk gebruik je binnen deze functie één keer APIerror. Dus dan zou ik ook gewoon `console.warn("Unable to collect country data. The network may be failing or the website may be down. Try again later.");` doen

Reply:
Omdat ik APIerror expres globaal heb gemaakt, is het m.i. beter het apart te houden.